labrad/node/__init__.py

```diff
@@ -503,95 +506,91 @@ class NodeServer(LabradServer):
+        # instances of the server.
         self.servers = {}
+
+        # Mapping from instance name to ServerProcess object. Instances are
+        # added to this mapping when they are successfully started and removed
+        # when they send a message indicating that the server process has
+        # stopped.
         self.instances = {}
-        self.initMessages(True)
+
+        # ServerProcess instances that will be notified when new servers connect
+        # to labrad. ServerProcesses use this information to determine when the
+        # subprocesses that they launch have successfully connected to labrad.
+        # We use a separate collection here than self.instances because
+        # ServerProcess instances need to receive the server connection
+        # notification in order to startup, but they are not added to
+        # self.instances until after they have started up successfully.
+        # TODO: ServerProcess instances should manage notifications on their
+        # own, since they have access to a labrad client.
+        self.instances_to_notify = weakref.WeakSet()
 
     @inlineCallbacks
     def initServer(self):
-        """Initialize this server."""
+        """Initialize this server after connecting to the labrad manager."""
         self.config = yield NodeConfig.create(self)
         self.refreshServers()
         self.autostart(None)
 
     def stopServer(self):
         """Stop this node by killing all subprocesses."""
-        self.initMessages(False)
         stoppages = [inst.stop() for inst in self.instances.values()]
         return defer.DeferredList(stoppages)
 
 
     # message handling
 
-    def initMessages(self, connect=True):
-        """Set up message dispatching."""
-        def f(receiver, signal):
-            try:
-                if connect:
-                    dispatcher.connect(receiver, signal)
-                else:
-                    dispatcher.disconnect(receiver, signal)
-            except dispatcher.DispatcherError as e:
-                method = 'connect' if connect else 'disconnect'
-                msg = ('Error while setting up message dispatching: '
-                       'method={}, receiver={}, signal={}.')
-                print(msg.format(method, receiver, signal), e)
-        # set up messages to be relayed out over LabRAD
-        messages = ['server_starting', 'server_started',
-                    'server_stopping', 'server_stopped',
-                    'status']
-        for message in messages:
-            f(self._relayMessage, message)
-        # set up message handlers for subprocess events
-        f(self.subprocessStarting, 'server_starting')
-        f(self.subprocessStarted, 'server_started')
-        f(self.subprocessStopping, 'server_stopping')
-        f(self.subprocessStopped, 'server_stopped')
-
-    def _relayMessage(self, signal, sender, **kw):
+    def on_subprocess_message(self, sender, message):
```

Comment:
A short doc here might be helpful. Also, AFAICT the only purpose of this is act as a `message_receiver` input to a `ServerProcess`, so maybe call it `subprocess_message_receiver` or `on_subprocess_message_receive`?